File: 0797-all-paths-from-source-to-target/0797-all-paths-from-source-to-target.py
class Solution:
    def allPathsSourceTarget(self, graph: List[List[int]]) -> List[List[int]]:
        ans = []
        path = []
        def dfs(cur):
            
            if cur == len(graph) - 1:
                path.append(cur)
                ans.append(list(path))
                path.pop()
                return
            
            path.append(cur)
            for node in graph[cur]:
                dfs(node)
            path.pop()
        
        dfs(0)
        
        return ans

File: 0797-all-paths-from-source-to-target/0797-all-paths-from-source-to-target.java
class Solution {
    public List<List<Integer>> allPathsSourceTarget(int[][] graph) {
        List<List<Integer>> ans = new ArrayList<>();
        
        dfs(graph, new ArrayList<Integer>(), new boolean[graph.length], 0, ans);
        
        return ans;
    }
    
    private void dfs(int[][] graph, List<Integer> cur, 
                     boolean[] onPath, int root,
                     List<List<Integer>> ans) {
        if (onPath[root]) return;
        if (root == graph.length - 1) {
            cur.add(root);
            ans.add(new ArrayList<>(cur));
            cur.remove(cur.size() - 1);
            return;
        }
        
        cur.add(root);
        onPath[root] = true;
        for (int n : graph[root]) {
            dfs(graph, cur, onPath, n, ans);
        }
        cur.remove(cur.size() - 1);
        onPath[root] = false;
        return;
    }
}